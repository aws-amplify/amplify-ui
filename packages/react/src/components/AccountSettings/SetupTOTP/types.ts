export interface TotpSecret {
  secretKey: string;
  qrCode: string;
}

export interface VerifyTotpStatus {
  isVerifying: boolean;
  errorMessage: string;
}

export interface ConfigureTOTPProps {
  /** TOTP issuer. Default to "AWSCognito" */
  totpIssuer?: string;
  /** TOTP username. Default to `user.username` */
  totpUsername?: string;
  /** callback once totp is successfully set up */
  onSuccess?: () => void;
  /** callback when there's an error */
  onError?: (error: Error) => void;
}
