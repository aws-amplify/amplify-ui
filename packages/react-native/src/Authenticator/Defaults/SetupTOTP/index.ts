export { default as SetupTOTP } from './SetupTOTP';
