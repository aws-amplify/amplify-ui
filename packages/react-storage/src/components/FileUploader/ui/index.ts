export { Container, ContainerProps } from './Container';
export { DropZone, DropZoneProps } from './DropZone';
export { FileList, FileListProps } from './FileList';
export { FileListHeader, FileListHeaderProps } from './FileListHeader';
export { FileListFooter, FileListFooterProps } from './FileListFooter';
export { FilePicker, FilePickerProps } from './FilePicker';
