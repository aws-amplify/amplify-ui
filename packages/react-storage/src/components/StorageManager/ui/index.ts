export { Container, ContainerProps } from './Container';
export {
  DropZone,
  DropZoneProps,
  FilePicker,
  FilePickerProps,
} from './DropZone';
export { FileList, FileListProps } from './FileList';
export { FileListHeader, FileListHeaderProps } from './FileListHeader';
