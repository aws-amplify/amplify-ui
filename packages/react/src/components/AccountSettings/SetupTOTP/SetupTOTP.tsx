import React from 'react';
import QRCode from 'qrcode';

import {
  AmplifyUser,
  getLogger,
  getTotpCodeURL,
  setupTOTP,
  translate,
  verifyTOTPToken,
} from '@aws-amplify/ui';

import { useAuth } from '../../../internal';
import { View, Flex } from '../../../primitives';
import { ComponentClassName } from '../types';
import {
  ConfirmationCodeField,
  CopyButton,
  ErrorMessage,
  QRCodeImage,
  SubmitButton,
} from './defaults';
import { ConfigureTOTPProps, TotpSecret, VerifyTotpStatus } from './types';
import { QR_CODE_DIMENSIONS } from './constants';

const logger = getLogger('Auth');

function SetupTOTP({
  totpIssuer = 'AWSCognito',
  totpUsername,
  onSuccess,
  onError,
}: ConfigureTOTPProps): JSX.Element | null {
  const [confirmationCode, setConfirmationCode] = React.useState<string>('');
  const [totpSecret, setTotpSecret] = React.useState<TotpSecret>(null);
  const [verifyTotpStatus, setVerifyTotpStatus] =
    React.useState<VerifyTotpStatus>({
      isVerifying: false,
      errorMessage: null,
    });

  const { user, isLoading } = useAuth();

  const generateQRCode = React.useCallback(
    async (currentUser: AmplifyUser): Promise<void> => {
      try {
        const secretKey = await setupTOTP(currentUser);
        const username = totpUsername ?? currentUser?.username;
        const totpCode = getTotpCodeURL(totpIssuer, username, secretKey);
        const qrCode = await QRCode.toDataURL(totpCode);

        setTotpSecret({ secretKey, qrCode });
      } catch (e) {
        logger.error(e);
      }
    },
    [totpIssuer, totpUsername]
  );

  React.useEffect(() => {
    if (user && !totpSecret) {
      generateQRCode(user);
    }
  }, [generateQRCode, totpSecret, user]);

  // translations
  const confirmText = translate('Confirm');
  const copyCodeText = translate('Copy Secret Code');

  // event handlers
  const handleChange = React.useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      event.preventDefault();
      const { value } = event.target;
      setConfirmationCode(value);
    },
    []
  );

  const runVerifyTotpToken = React.useCallback(
    async ({ user, code }: { user: AmplifyUser; code: string }) => {
      setVerifyTotpStatus({ isVerifying: true, errorMessage: null });

      try {
        await verifyTOTPToken({ user, code });

        setVerifyTotpStatus({ isVerifying: false, errorMessage: null });

        onSuccess?.();
      } catch (e) {
        const error = e as Error;

        setVerifyTotpStatus({
          isVerifying: false,
          errorMessage: error.message,
        });

        onError?.(error);
      }
    },
    [onError, onSuccess]
  );

  const handleSubmit = React.useCallback(
    (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();
      runVerifyTotpToken({ user, code: confirmationCode });
    },
    [user, confirmationCode, runVerifyTotpToken]
  );

  const handleCopy = React.useCallback(() => {
    navigator.clipboard.writeText(totpSecret.secretKey);
  }, [totpSecret]);

  // return null if Auth.getCurrentAuthenticatedUser is still in progress
  if (isLoading) {
    return null;
  }

  // return null if user isn't authenticated in the first place
  if (!user) {
    logger.warn('<SetupTOTP /> requires user to be authenticated.');
    return null;
  }

  const { isVerifying, errorMessage } = verifyTotpStatus;

  return (
    <View
      as="form"
      className={ComponentClassName.SetupTOTP}
      onSubmit={handleSubmit}
    >
      <Flex direction="column" alignItems="center">
        {totpSecret?.qrCode ? (
          <QRCodeImage
            src={totpSecret?.qrCode}
            alt="qr code"
            {...QR_CODE_DIMENSIONS}
          />
        ) : null}
        <CopyButton
          isDisabled={isVerifying}
          alignSelf="stretch"
          onClick={handleCopy}
        >
          {copyCodeText}
        </CopyButton>
        <ConfirmationCodeField
          alignSelf="stretch"
          isRequired
          label="Confirmation Code"
          name="code"
          onChange={handleChange}
          placeholder="Code"
          value={confirmationCode}
          isDisabled={isVerifying}
        />

        <SubmitButton
          type="submit"
          variation="primary"
          isDisabled={isVerifying || confirmationCode.length === 0}
          isFullWidth
        >
          {confirmText}
        </SubmitButton>
        {errorMessage ? (
          <ErrorMessage width="100%">{errorMessage}</ErrorMessage>
        ) : null}
      </Flex>
    </View>
  );
}

export default SetupTOTP;
