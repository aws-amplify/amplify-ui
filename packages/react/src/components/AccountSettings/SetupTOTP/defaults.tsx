import React from 'react';

import { Alert, Button, Image, TextField } from '../../../primitives';
import {
  ButtonComponent,
  ErrorMessageComponent,
  ImageComponent,
  SubmitButtonComponent,
  TextFieldComponent,
} from '../types';

export const ConfirmationCodeField: TextFieldComponent = (props) => (
  <TextField {...props} />
);

export const QRCodeImage: ImageComponent = Image;

export const CopyButton: ButtonComponent = Button;

export const SubmitButton: SubmitButtonComponent = (props) => (
  <Button type="submit" {...props} />
);

export const ErrorMessage: ErrorMessageComponent = (props) => {
  return <Alert variation="error" {...props} />;
};
