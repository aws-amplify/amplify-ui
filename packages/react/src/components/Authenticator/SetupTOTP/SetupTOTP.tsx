import QRCode from 'qrcode';
import * as React from 'react';

import { Auth, Logger } from 'aws-amplify';
import {
  CognitoUserAmplify,
  getActorState,
  SignInState,
  translate,
} from '@aws-amplify/ui';

import { Flex } from '../../../primitives/Flex';
import { Heading } from '../../../primitives/Heading';
import { useAuthenticator } from '../hooks/useAuthenticator';
import { useCustomComponents } from '../hooks/useCustomComponents';
import { useFormHandlers } from '../hooks/useFormHandlers';
import { ConfirmSignInFooter } from '../shared/ConfirmSignInFooter';
import { RemoteErrorMessage } from '../shared/RemoteErrorMessage';
import { FormFields } from '../shared/FormFields';

const logger = new Logger('SetupTOTP-logger');

export const SetupTOTP = (): JSX.Element => {
  // TODO: handle `formOverrides` outside `useAuthenticator`
  const { _state, isPending } = useAuthenticator((context) => [
    context.isPending,
  ]);
  const { handleChange, handleSubmit } = useFormHandlers();

  const {
    components: {
      SetupTOTP: { Header = SetupTOTP.Header, Footer = SetupTOTP.Footer },
    },
  } = useCustomComponents();

  const [isLoading, setIsLoading] = React.useState<boolean>(true);
  const [qrCode, setQrCode] = React.useState<string>();
  const [copyTextLabel, setCopyTextLabel] = React.useState<string>('COPY');
  const [secretKey, setSecretKey] = React.useState<string>('');

  // `user` hasn't been set on the top-level state yet, so it's only available from the signIn actor
  const actorState = getActorState(_state) as SignInState;

  const { formFields, user } = actorState.context;
  const { totpIssuer = 'AWSCognito', totpUsername = user.username } =
    formFields?.setupTOTP?.QR;

  const generateQRCode = React.useCallback(
    async (currentUser: CognitoUserAmplify): Promise<void> => {
      try {
        const newSecretKey = await Auth.setupTOTP(currentUser);
        setSecretKey(newSecretKey);
        const totpCode = `otpauth://totp/${totpIssuer}:${totpUsername}?secret=${newSecretKey}&issuer=${totpIssuer}`;
        const qrCodeImageSource = await QRCode.toDataURL(totpCode);

        setQrCode(qrCodeImageSource);
      } catch (error) {
        logger.error(error);
      } finally {
        setIsLoading(false);
      }
    },
    [totpIssuer, totpUsername]
  );

  React.useEffect(() => {
    if (!user) return;

    generateQRCode(user);
  }, [generateQRCode, user]);

  const copyText = (): void => {
    navigator.clipboard.writeText(secretKey);
    setCopyTextLabel(translate('COPIED'));
  };

  return (
    <form
      data-amplify-form=""
      data-amplify-authenticator-setup-totp=""
      method="post"
      onChange={handleChange}
      onSubmit={handleSubmit}
    >
      <fieldset
        style={{ display: 'flex', flexDirection: 'column' }}
        className="amplify-flex"
        disabled={isPending}
      >
        <Header />

        <Flex direction="column">
          {/* TODO: Add spinner here instead of loading text... */}
          {isLoading ? (
            <p>{translate('Loading')}&hellip;</p>
          ) : (
            <img
              data-amplify-qrcode
              src={qrCode}
              alt="qr code"
              width="228"
              height="228"
            />
          )}
          <Flex data-amplify-copy>
            <div>{secretKey}</div>
            <Flex data-amplify-copy-svg onClick={copyText}>
              <div data-amplify-copy-tooltip>{copyTextLabel}</div>
              <svg
                width="24"
                height="24"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path d="M16 1H4C2.9 1 2 1.9 2 3V17H4V3H16V1ZM15 5H8C6.9 5 6.01 5.9 6.01 7L6 21C6 22.1 6.89 23 7.99 23H19C20.1 23 21 22.1 21 21V11L15 5ZM8 21V7H14V12H19V21H8Z" />
              </svg>
            </Flex>
          </Flex>
          <FormFields route="setupTOTP" />
          <RemoteErrorMessage />
        </Flex>

        <ConfirmSignInFooter />
        <Footer />
      </fieldset>
    </form>
  );
};

SetupTOTP.Header = () => {
  return <Heading level={3}>{translate('Setup TOTP')}</Heading>;
};

SetupTOTP.Footer = (): JSX.Element => null;
