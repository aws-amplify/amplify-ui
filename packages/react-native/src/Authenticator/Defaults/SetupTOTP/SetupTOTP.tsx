import React, { useMemo } from 'react';

import { authenticatorTextUtil } from '@aws-amplify/ui';

import { Label } from '../../../primitives';

import {
  DefaultContent,
  DefaultFooter,
  DefaultTextFormFields,
  DefaultHeader,
} from '../../common';
import { useFieldValues } from '../../hooks';

import { DefaultSetupTOTPProps } from '../types';
import { styles } from './styles';

const COMPONENT_NAME = 'SetupTOTP';

const {
  getBackToSignInText,
  getConfirmingText,
  getConfirmText,
  getSetupTOTPText,
  getSetupTOTPInstructionsText,
} = authenticatorTextUtil;

const SetupTOTP = ({
  fields,
  handleBlur,
  handleChange,
  handleSubmit,
  isPending,
  toSignIn,
  totpSecretCode,
  validationErrors,
  ...rest
}: DefaultSetupTOTPProps): JSX.Element => {
  const {
    disableFormSubmit: disabled,
    fields: fieldsWithHandlers,
    fieldValidationErrors,
    handleFormSubmit,
  } = useFieldValues({
    componentName: COMPONENT_NAME,
    fields,
    handleBlur,
    handleChange,
    handleSubmit,
    validationErrors,
  });

  const headerText = getSetupTOTPText();
  const primaryButtonText = isPending ? getConfirmingText() : getConfirmText();
  const secondaryButtonText = getBackToSignInText();

  const body = (
    <>
      <Label style={styles.secretKeyText}>
        {getSetupTOTPInstructionsText()}
      </Label>
      <Label selectable style={styles.secretKeyText}>
        {totpSecretCode}
      </Label>
    </>
  );

  const buttons = useMemo(
    () => ({
      primary: {
        children: primaryButtonText,
        disabled,
        onPress: handleFormSubmit,
      },
      links: [{ children: secondaryButtonText, onPress: toSignIn }],
    }),
    [
      disabled,
      handleFormSubmit,
      primaryButtonText,
      secondaryButtonText,
      toSignIn,
    ]
  );

  return (
    <DefaultContent
      {...rest}
      body={body}
      buttons={buttons}
      headerText={headerText}
      fields={fieldsWithHandlers}
      isPending={isPending}
      validationErrors={fieldValidationErrors}
    />
  );
};

SetupTOTP.Footer = DefaultFooter;
SetupTOTP.FormFields = DefaultTextFormFields;
SetupTOTP.Header = DefaultHeader;

SetupTOTP.displayName = COMPONENT_NAME;
export default SetupTOTP;
