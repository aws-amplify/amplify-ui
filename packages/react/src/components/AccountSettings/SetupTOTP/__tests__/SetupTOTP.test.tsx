import React from 'react';
import {
  render,
  screen,
  fireEvent,
  waitFor,
  act,
} from '@testing-library/react';

import * as UIModule from '@aws-amplify/ui';

import SetupTOTP from '../SetupTOTP';

const user = { username: 'testuser' } as unknown as UIModule.AmplifyUser;
jest.mock('../../../../internal', () => ({
  useAuth: () => ({
    user,
    isLoading: false,
  }),
}));

const setupTOTPSpy = jest.spyOn(UIModule, 'setupTOTP');
const verifyTOTPToken = jest.spyOn(UIModule, 'verifyTOTPToken');

describe('SetupTOTP', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders as expected', () => {
    setupTOTPSpy.mockResolvedValue('secretcode');
    const { container } = render(<SetupTOTP />);
    expect(container).toMatchSnapshot();
  });

  it('calls setupTOTP with expected arguments', async () => {
    setupTOTPSpy.mockResolvedValue('secretcode');

    render(<SetupTOTP />);

    await screen.findByAltText('qr code');
    expect(setupTOTPSpy).toHaveBeenCalledWith(user);
  });

  it('calls onSuccess on successful totp verification', async () => {
    setupTOTPSpy.mockResolvedValue('secretcode');
    verifyTOTPToken.mockResolvedValue();

    const onSuccess = jest.fn();

    await act(async () => {
      render(<SetupTOTP onSuccess={onSuccess} />);
    });

    const submitButton = await screen.findByRole('button', {
      name: 'Confirm',
    });
    fireEvent.submit(submitButton);

    // submit handling is async, wait for onSuccess to be called
    // https://testing-library.com/docs/dom-testing-library/api-async/#waitfor
    await waitFor(() => expect(onSuccess).toBeCalledTimes(1));
  });

  it('calls onError after unsuccessful submit', async () => {
    setupTOTPSpy.mockResolvedValue('secretCode');
    verifyTOTPToken.mockRejectedValue(new Error('Mock Error'));

    const onError = jest.fn();
    await act(async () => {
      render(<SetupTOTP onError={onError} />);
    });
    const submitButton = await screen.findByRole('button', {
      name: 'Confirm',
    });

    fireEvent.submit(submitButton);

    // submit handling is async, wait for onError to be called
    await waitFor(() => expect(onError).toBeCalledTimes(1));
  });

  it('displays an error message on unsuccessful submit', async () => {
    setupTOTPSpy.mockResolvedValue('secretCode');
    verifyTOTPToken.mockRejectedValue(new Error('Mock Error'));

    const onError = jest.fn();

    await act(async () => {
      render(<SetupTOTP onError={onError} />);
    });

    const submitButton = await screen.findByRole('button', {
      name: 'Confirm',
    });

    fireEvent.submit(submitButton);

    // submit handling is async, wait for error to be displayed
    expect(await screen.findByText('Mock Error')).toBeDefined();
  });
});
