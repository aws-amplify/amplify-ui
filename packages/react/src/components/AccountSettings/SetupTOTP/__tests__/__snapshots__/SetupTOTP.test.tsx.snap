// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetupTOTP renders as expected 1`] = `
<div>
  <form
    class="amplify-accountsettings-setuptotp"
  >
    <div
      class="amplify-flex"
      style="flex-direction: column; align-items: center;"
    >
      <button
        class="amplify-button amplify-field-group__control"
        data-fullwidth="false"
        style="align-self: stretch;"
        type="button"
      >
        Copy Secret Code
      </button>
      <div
        class="amplify-flex amplify-field amplify-textfield"
        style="align-self: stretch;"
      >
        <label
          class="amplify-label"
          for="amplify-id-0"
        >
          Confirmation Code
        </label>
        <div
          class="amplify-flex amplify-field-group amplify-field-group--horizontal"
          data-orientation="horizontal"
        >
          <div
            class="amplify-field-group__field-wrapper amplify-field-group__field-wrapper--horizontal"
            data-orientation="horizontal"
          >
            <input
              aria-invalid="false"
              class="amplify-input amplify-field-group__control"
              id="amplify-id-0"
              name="code"
              placeholder="Code"
              required=""
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <button
        class="amplify-button amplify-field-group__control amplify-button--primary amplify-button--disabled amplify-button--fullwidth"
        data-fullwidth="true"
        data-variation="primary"
        disabled=""
        type="submit"
      >
        Confirm
      </button>
    </div>
  </form>
</div>
`;
